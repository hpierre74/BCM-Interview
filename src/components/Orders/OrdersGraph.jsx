import React from "react";
import {
  LineChart,
  Line,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
  Brush
} from "recharts";
import moment from "moment";

const OrdersGraph = props => {

  // Unnecessary loop to improve user dates readability
  // Workaround / additional instructions about date displaying needed
  const groupDates = () => {
    const newData = [];
    Object.values(props.data).map((order, index) => {
      order.date = moment(order.datetime).format("DD/MM/YYYY");
      return newData.push(order);
    });
    return newData;
  };

  return (
    <div>
      <h1> {props.title} </h1>
      <LineChart
        width={window.innerWidth / 2.5}
        height={300}
        data={groupDates()}
        margin={{ top: 5, right: 30, left: 20, bottom: 5 }}
      >
        <XAxis dataKey="date" />
        <YAxis />
        <CartesianGrid strokeDasharray="3 3" />
        <Tooltip />
        <Legend />
        <Brush dataKey="date" height={30} stroke={props.primary} />
        <Line
          type="monotone"
          dataKey="price"
          stroke={props.primary}
          activeDot={{ r: 8 }}
        />
        <Line type="monotone" dataKey="datetime" stroke={props.secondary} />
      </LineChart>
    </div>
  );
};

export default OrdersGraph;
