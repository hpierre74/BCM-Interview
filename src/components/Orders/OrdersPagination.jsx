import React from "react";
import { Pagination, PaginationItem, PaginationLink } from "reactstrap";

const OrdersPagination = props => {
  const { orders, page, paginate, isVisible, displayPagination } = props;

  const renderPaginations = () => {
    return Object.values(orders).map((order, index) => {
      return (
        <PaginationItem key={index}>
          <PaginationLink onClick={e => paginate(e, index)} value={index}>
            {index}
          </PaginationLink>
        </PaginationItem>
      );
    });
  };
  const styles = {
    justifyContent: "flex-end",
    maxWidth: "100%",
    display: "flex",
    flexFlow: "row wrap"
  };
  return (
    <Pagination style={styles}>
      <PaginationItem>
        <PaginationLink onClick={e => paginate(e, page - 1)} previous />
      </PaginationItem>
      {isVisible ? (
        renderPaginations()
      ) : (
        <PaginationItem>
          <PaginationLink> {page} </PaginationLink>
        </PaginationItem>
      )}
      <PaginationItem>
        <PaginationLink onClick={e => displayPagination(e)}>
          {" "}
          ...{" "}
        </PaginationLink>
      </PaginationItem>
      <PaginationItem>
        <PaginationLink next onClick={e => paginate(e, page + 1)} />
      </PaginationItem>
    </Pagination>
  );
};
export default OrdersPagination;
