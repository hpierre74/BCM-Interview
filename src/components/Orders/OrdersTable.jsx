import React from "react";
import { Table } from "reactstrap";
import moment from "moment";

const OrdersTable = props => {
  const { orders } = props;
  const renderOrders = () => {
    return Object.values(orders).map((order, index) => {
      return (
        <tr key={index}>
          <td> {moment(order.datetime).format("DD/MM/YYYY HH:mm:ss")}</td>
          <td> {order.type} </td>
          <td> {order.volume} </td>
          <td> {order.price + " " + order.currency} </td>
        </tr>
      );
    });
  };
  return (
    <div>
      <Table striped className="OrdersLayout">
        <thead>
          <tr className="OrdersLayout-header">
            <th>Date</th>
            <th>Order</th>
            <th>Volume</th>
            <th>Price</th>
          </tr>
        </thead>
        <tbody>{renderOrders()}</tbody>
      </Table>
    </div>
  );
};

export default OrdersTable;
