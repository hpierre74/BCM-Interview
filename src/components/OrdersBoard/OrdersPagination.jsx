import React from "react";
import { Pagination, PaginationItem, PaginationLink, Input } from "reactstrap";
import PropTypes from 'prop-types';

const OrdersPagination = props => {
  const { page, paginate } = props;

  const styles = {
    justifyContent: "flex-end",
    maxWidth: "100%",
    display: "flex",
    flexFlow: "row wrap"
  };
  return (
    <Pagination style={styles}>
      <PaginationItem>
        <PaginationLink onClick={e => paginate(e, page - 1)} previous />
      </PaginationItem>
      <PaginationItem>
        <PaginationLink> {page} </PaginationLink>
      </PaginationItem>
      <PaginationItem>
        <PaginationLink next onClick={e => paginate(e, page + 1)} />
      </PaginationItem>
      <Input
        onChange={e => paginate(e, e.target.value)}
        style={{ width: "100px" }}
        type="number"
        name="pickPage"
        id="pickPage"
        placeholder="go to"
      />
    </Pagination>
  );
};

OrdersPagination.propTypes= {
  page: PropTypes.number.isRequired,
  paginate: PropTypes.func.isRequired,
}

export default OrdersPagination;
