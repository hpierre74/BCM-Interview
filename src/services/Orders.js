import axios from "axios";

const orderInstance = axios.create({
  baseURL: "http://localhost:3001/orders/"
});

class Orders {
  static getOrders(type, cb, page) {
    return page
      ? orderInstance
          .get(type + "?page=" + page)
          .then(res => {
            return page
              ? cb(res.data.orders, type, true)
              : cb(res.data.orders, type, false);
          })
          .catch(err => console.log(err))
      : orderInstance
          .get(type)
          .then(res => {
            return cb(res.data.orders, type);
          })
          .catch(err => console.log(err));
  }
}
export default Orders;
