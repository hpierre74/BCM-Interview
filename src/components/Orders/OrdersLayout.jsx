import React, { Component } from "react";
import { Container, Row, Col } from "reactstrap";
import axios from "axios";
import OrdersTable from "./OrdersTable";
import OrdersPagination from "./OrdersPagination";
import OrdersGraph from "./OrdersGraph";

class OrdersLayout extends Component {
  constructor(context, props) {
    super(context, props);
    this.state = {
      page: 0,
      allOrders: [],
      isPaginationVisible: false,
      chunkedOrders: [],
      sellOrders: [],
      buyOrders: []
    };
    this.req = axios.create({
      baseURL: "http://localhost:3001/orders/"
    });
    //I skipped pagination requests as it wasn't technically hard to implement
    //I would normally use the pagination or set data in session/cache/cookies
    //to avoid multiple requests
    //this.paging = "?page=" + this.state.page;
  }

  componentWillMount() {
    this.setOrders();
  }

  setOrders() {
    this.getOrders("ALL", "allOrders");
    this.getOrders("SELL", "sellOrders");
    this.getOrders("BUY", "buyOrders");
  }

  getOrders(type, state) {
    this.req
      .get(type)
      .then(res => {
        return this.setState(
          {
            [state]: res.data.orders
          },
          () => {
            return state === "allOrders"
              ? this.chunkOrders(res.data.orders)
              : "";
          }
        );
      })
      .catch(err => console.log(err));
  }

  //Set the current Table page value
  paginate(e, page) {
    e.preventDefault();
    this.setState({ page: page });
  }

  //toggle Pagination links visibility
  displayPagination(e) {
    e.preventDefault();
    this.setState({ isPaginationVisible: !this.state.isPaginationVisible });
  }

  //Change the array to multiple ones with length relying on given chunk_size param
  chunkArray(myArray, chunk_size) {
    let results = [];
    while (myArray.length) {
      results.push(myArray.splice(0, chunk_size));
    }
    return results;
  }

  chunkOrders(orders) {
    return this.setState({ chunkedOrders: this.chunkArray(orders, 10) });
  }

  render() {
    return (
      <Container fluid>
        <Row>
          <Col style={{ backgroundColor: "#434449" }} sm="2" md="2" xs="2">
            Orders...
          </Col>
          <Col>
            <Row>
              <Col>
                <OrdersTable
                  orders={
                    this.state.chunkedOrders[this.state.page]
                      ? this.state.chunkedOrders[this.state.page]
                      : []
                  }
                />
                <OrdersPagination
                  isVisible={this.state.isPaginationVisible}
                  displayPagination={this.displayPagination.bind(this)}
                  page={this.state.page}
                  paginate={this.paginate.bind(this)}
                  orders={this.state.chunkedOrders}
                />
              </Col>
            </Row>
            <Row>
              <Col style={{ display: "flex", flexFlow: "row wrap" }}>
                <OrdersGraph
                  title="Selling Orders"
                  primary="#FF5733"
                  secondary="#FF8A33"
                  data={this.state.sellOrders}
                />
                <OrdersGraph
                  title="Buying Orders"
                  primary="#8884d8"
                  secondary="#82ca9d"
                  data={this.state.buyOrders}
                />
              </Col>
            </Row>
          </Col>
        </Row>
      </Container>
    );
  }
}

export default OrdersLayout;
