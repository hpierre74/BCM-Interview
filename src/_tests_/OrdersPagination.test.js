import OrdersPagination from "../components/OrdersBoard/OrdersPagination";
import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import Orders from "../services/orders";


let paginate = () => {
    return;
}
let page = 0;


describe("OrdersPagination component renders the children correctly", () => {
  it("renders correctly", () => {
    const rendered = renderer.create(
      <OrdersPagination paginate={paginate} page={page} />
    );
    expect(rendered.toJSON()).toMatchSnapshot();
  });
});
