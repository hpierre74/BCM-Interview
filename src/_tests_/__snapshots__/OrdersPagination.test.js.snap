// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersPagination component renders the children correctly renders correctly 1`] = `
<ul
  className="pagination"
  style={
    Object {
      "display": "flex",
      "flexFlow": "row wrap",
      "justifyContent": "flex-end",
      "maxWidth": "100%",
    }
  }
>
  <li
    className="page-item"
  >
    <a
      aria-label="Previous"
      className="page-link"
      onClick={[Function]}
    >
      <span
        aria-hidden="true"
      >
        «
      </span>
      <span
        className="sr-only"
      >
        Previous
      </span>
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      aria-label={undefined}
      className="page-link"
    >
       
      0
       
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      aria-label="Next"
      className="page-link"
      onClick={[Function]}
    >
      <span
        aria-hidden="true"
      >
        »
      </span>
      <span
        className="sr-only"
      >
        Next
      </span>
    </a>
  </li>
  <input
    className="form-control"
    id="pickPage"
    name="pickPage"
    onChange={[Function]}
    placeholder="go to"
    style={
      Object {
        "width": "100px",
      }
    }
    type="number"
  />
</ul>
`;
